### DISCUSSION TITLE
Relation notifications

### DISCUSSION THREAD

**Discussion by bjarnef** [umbraco/Umbraco-CMS#13377] [Features and ideas]
Currently there a lot of different notifications to hook into.
https://our.umbraco.com/documentation/Reference/Notifications/

However there isn't any relation specific notifications, which I think would be useful when relations are created or deleted, e.g.
- RelationSavingNotification
- RelationSavedNotification
- RelationDeletingNotification
- RelationDeletedNotification

For example this could be use to "sync" a two pickers between Page A and B, where picker on Page A select page of type B, and picker on Page B select page of type A. In that way the property value could be updated when the relation change.

**Comment by bergmania** [comment]
Hey @bjarnef.

I agree, but to my knowledge these notifications already exists.. https://github.com/umbraco/Umbraco-CMS/blob/v11/contrib/src/Umbraco.Core/Services/RelationService.cs

So the question is, if it is "just" missing documentation :) ?

**Reply by bjarnef** [reply]
@bergmania yes, it looks like some missing documentation like `ContentCacheRefresherNotification` isn't documented either.
https://github.com/umbraco/UmbracoDocs/issues/4536

I will try these. Besides that I noticed the `UmbracoEntityReference` didn't work as expected with a custom relation type as it doesn't removed previous relations, so it tries to insert new records which already exists. Currently it only works with `umbDocument` and `umbMedia` defined in `AutomaticRelationTypes` constant.
https://github.com/umbraco/Umbraco-CMS/issues/13364

**Reply by bjarnef** [reply]
@bergmania I tried with `RelationDeletingNotification` but if check for deleted reference and update the property on the other one (the one picked) it seems I would get an infinite loop, because it always remove the relations first, before the new ones is created. Not sure if it could only remove relations which are removed, e.g. if A, B and C is picked and C is removed... it seems to remove all and create A and B relations again.
https://github.com/umbraco/Umbraco-CMS/issues/13364#issuecomment-1308577311

**Reply by bjarnef** [reply]
Am I correct that `RelationDeletingNotification` and `RelationDeletedNotification` notifications are triggered for all items in picker, e.g. A, B and C.
https://github.com/umbraco/Umbraco-CMS/blob/b69afe81f3f6fcd37480b3b0295a62af44ede245/src/Umbraco.Infrastructure/Persistence/Repositories/Implement/ContentRepositoryBase.cs#L1045-L1046

and afterwards the new references are saved based on implementation of `UmbracoEntityReference` in `GetReferences()`:
https://github.com/umbraco/Umbraco-CMS/blob/b69afe81f3f6fcd37480b3b0295a62af44ede245/src/Umbraco.Infrastructure/Persistence/Repositories/Implement/ContentRepositoryBase.cs#L1064-L1085

I think it would be useful it only triggered a notification for the ones deleted/removed. In my use-case I have a picker on a Category node to select courses, e.g. Course 1, Course 2 and Course 3.

Even I only remove Course 2, all three are triggering the `RelationDeletedNotification`.
So if I wants to "sync" a reference in a Category picker on the Course nodes, it would also remove this Category from Course 1 and Course 3 even though only Course 2 was was removed from the Category. I could of course handle the `RelationSavingNotification` or `RelationSavedNotification` to update the values on Course 1 and 3, but it would make on unnecessary save of Course 1 and 3, when only Course 2 is the one expected to update (since this was removed from the course picker on the category).
